ees_sharepoint/sync_sharepoint.py

```diff
@@ -0,0 +1,614 @@
+from .checkpointing import Checkpoint
+from .usergroup_permissions import Permissions
+from .utils import encode, extract, split_documents_into_equal_chunks, split_list_into_buckets
+
+IDS_PATH = os.path.join(os.path.dirname(__file__), "doc_id.json")
+
+SITE = "site"
+LIST = "list"
+ITEM = "item"
+SITES = "sites"
+LISTS = "lists"
+LIST_ITEMS = "list_items"
+DRIVE_ITEMS = "drive_items"
+
+
+def get_results(logger, response, entity_name):
+    """Attempts to fetch results from a Sharepoint Server response
+    :param response: response from the sharepoint client
+    :param entity_name: entity name whether it is SITES, LISTS, LIST_ITEMS OR DRIVE_ITEMS
+    Returns:
+        Parsed response
+    """
+    if not response:
+        logger.error(f"Empty response when fetching {entity_name}")
+        return None
+
+    if entity_name == "attachment" and not response.get("d", {}).get("results"):
+        logger.info("Failed to fetch attachment")
+        return None
+    return response.get("d", {}).get("results")
+
+
+class SyncSharepoint:
+    """This class allows syncing objects from the SharePoint Server."""
+
+    def __init__(
+            self,
+            config,
+            logger,
+            workplace_search_client,
+            sharepoint_client,
+            start_time,
+            end_time,
+            queue,
+    ):
+        self.config = config
+        self.logger = logger
+        self.workplace_search_client = workplace_search_client
+        self.sharepoint_client = sharepoint_client
+
+        self.ws_source = config.get_value("workplace_search.source_id")
+        self.objects = config.get_value("objects")
+        self.site_collections = config.get_value("sharepoint.site_collections")
+        self.enable_permission = config.get_value("enable_document_permission")
+        self.start_time = start_time
+        self.end_time = end_time
+        self.sharepoint_thread_count = config.get_value("sharepoint_sync_thread_count")
+        self.mapping_sheet_path = config.get_value("sharepoint_workplace_user_mapping")
+        self.sharepoint_host = config.get_value("sharepoint.host_url")
+        self.checkpoint = Checkpoint(config, logger)
+        self.permissions = Permissions(
+            self.sharepoint_client, self.workplace_search_client, logger
+        )
+        self.queue = queue
+
+    def get_schema_fields(self, document_name):
+        """returns the schema of all the include_fields or exclude_fields specified in the configuration file.
+        :param document_name: document name from SITES, LISTS, LIST_ITEMS OR DRIVE_ITEMS
+        Returns:
+            schema: included and excluded fields schema
+        """
+        fields = self.objects.get(document_name)
+        adapter_schema = adapter.DEFAULT_SCHEMA[document_name]
+        field_id = adapter_schema["id"]
+        if fields:
+            include_fields = fields.get("include_fields")
+            exclude_fields = fields.get("exclude_fields")
```

Comment:
What if both are provided? Can we throw an error, if so?

Reply:
Hi @artem-shelkovnikov, In case of overlap between `include_fields` and `exclude_fields`, priority is given to `include_fields`.